### ISSUE TITLE
[BLOCKING] /root/repo/xgboost/src/objective/regression_obj.cu:65: Check failed: preds.Size() == info.labels_.Size() (904 vs. 3616) labels are not correctly providedpreds.size=904, label.size=3616

### ISSUE THREAD

**Issue by pseudotensor** [dmlc/xgboost#4163]
1dac5e24108890c946c5a4673f0d1fa4d6125943 on fresh build off dmlc

[failxgbregression.zip](https://github.com/dmlc/xgboost/files/2881889/failxgbregression.zip)

@hcho3 seems to be **major** regression after recent fixes.  Seems all regression problems have the issue.

**Comment by pseudotensor** [comment]
I get this on all or most (not sure) regression problems

**Comment by hcho3** [comment]
Can you try running the same thing with 0ff84d950ed78df9941ccbaec3ac08e15159f81a ? #4147 might have introduced the bug.

**Comment by trivialfis** [comment]
@pseudotensor It might be my fault on #4147 .  But I ran your script fine:

```
...
[535]	validation_0-rmse:0.14538
[536]	validation_0-rmse:0.145462
[537]	validation_0-rmse:0.145495
[538]	validation_0-rmse:0.145489
[539]	validation_0-rmse:0.145514
[540]	validation_0-rmse:0.145533
[541]	validation_0-rmse:0.145551
[542]	validation_0-rmse:0.145574
[543]	validation_0-rmse:0.145422
[544]	validation_0-rmse:0.145423
[545]	validation_0-rmse:0.145428
Stopping. Best iteration:
[445]	validation_0-rmse:0.145298
```

**Comment by trivialfis** [comment]
@hcho3 If confirmed that PR is to blame, I will summit a PR to revert and reopen the closed bug since I can't reproduce.

**Comment by pseudotensor** [comment]
Ok, I will check latest hash.

**Comment by hcho3** [comment]
@pseudotensor @trivialfis I actually reproduced the bug on my Linux machine with the latest master. On the other hand, with commit hash 0ff84d950ed78df9941ccbaec3ac08e15159f81a, the example runs fine.

**Comment by hcho3** [comment]
Let's see if there is a quick fix for this

**Comment by hcho3** [comment]
Okay, found the issue: somehow the variable `num` in
https://github.com/dmlc/xgboost/blob/c8c472f39aa08d5b8f180076a429f09317451cc7/src/data/data.cc#L140-L141
is not correct. The `num` should have been 904 in the given example but now is being set to 3616.

**Comment by trivialfis** [comment]
@hcho3 Thanks!  Seems my master branch is one commit behind.  Sorry about that.  Looking into this.

**Comment by hcho3** [comment]
@trivialfis I actually have a one-liner fix. Trying to fix the test `test_np_view` now

**Comment by trivialfis** [comment]
@hcho3 Yap.  the `length` from Python called should be divided by `data.dtype.itemsize`:
https://github.com/dmlc/xgboost/blob/c8c472f39aa08d5b8f180076a429f09317451cc7/python-package/xgboost/core.py#L225

**Comment by hcho3** [comment]
@trivialfis It turns out that `len(data.base)` may mean different things. Sometimes it shows the number of bytes, and sometimes it shows the number of elements. Rather than dealing with NumPy internals, I'd prefer to simply make copies.

**Comment by trivialfis** [comment]
@hcho3 Okay, I will revert it tomorrow.  Really need some sleep now. It's 6AM at my place...

**Comment by hcho3** [comment]
@trivialfis I'll take care of it. Good night

**Comment by hcho3** [comment]
@pseudotensor Fix is available at #4165. Thanks for reporting the bug.

**Comment by pseudotensor** [comment]
Thanks!
